test/CMakeLists.txt

```diff
@@ -2,9 +2,35 @@ include(CheckTest)
 
 find_package(Check REQUIRED)
 
-set(CHECK_LIBS ${CMAKE_THREAD_LIBS_INIT})
+set(CHECK_LIBS acf ${CMAKE_THREAD_LIBS_INIT})
 
-CHECK_TEST(check_an_array LIBS ${CHECK_LIBS} SRC ${SRC} check_an_array.c)
-CHECK_TEST(check_charset LIBS ${CHECK_LIBS} SRC ${SRC} check_charset.c)
-CHECK_TEST(check_bitmap LIBS ${CHECK_LIBS} SRC ${SRC} check_bitmap.c)
-CHECK_TEST(check_an_array_cpp LIBS ${CHECK_LIBS} SRC ${SRC} check_an_array_cpp.cc)
\ No newline at end of file
+if (${COVERAGE})
+    message(STATUS "Building test suite with coverage information")
+    if(NOT "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+        message(FATAL_ERROR "Code coverage only works with GCC")
+    endif()
+
+    # enable profiling
+    set(CMAKE_CXX_FLAGS "--coverage -g -O0 -fprofile-arcs -ftest-coverage")
+
+    # from https://github.com/RWTH-HPC/CMake-codecov/blob/master/cmake/FindGcov.cmake
+    get_filename_component(COMPILER_PATH "${CMAKE_CXX_COMPILER}" PATH)
+    string(REGEX MATCH "^[0-9]+" GCC_VERSION "${CMAKE_CXX_COMPILER_VERSION}")
+    find_program(GCOV_BIN NAMES gcov-${GCC_VERSION} gcov HINTS ${COMPILER_PATH})
+
+    # add target to collect coverage information and generate HTML file
+    # (filter script from https://stackoverflow.com/a/43726240/266378)
+    add_custom_target(lcov_html
+        COMMAND lcov --directory . --capture --output-file json.info --gcov-tool ${GCOV_BIN} --rc lcov_branch_coverage=1
+        COMMAND lcov -e json.info ${CMAKE_SOURCE_DIR}/src/json.hpp --output-file json.info.filtered --rc lcov_branch_coverage=1
+        COMMAND ${CMAKE_SOURCE_DIR}/test/thirdparty/imapdl/filterbr.py json.info.filtered > json.info.filtered.noexcept
+        COMMAND genhtml --title "JSON for Modern C++" --legend --demangle-cpp --output-directory html --show-details --branch-coverage json.info.filtered.noexcept
+        COMMENT "Generating HTML report test/html/index.html"
+    )
+endif()
+
+
+CHECK_TEST(check_an_array LIBS ${CHECK_LIBS} SRC check_an_array.c)
```

Comment:
is this set of changes necessary? linking directly against the .o makes it easier to access hidden symbols.